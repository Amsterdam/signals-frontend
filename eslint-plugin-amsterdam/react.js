module.exports = {
  extends: ['plugin:react/all'],
  plugins: ['jsx-a11y', 'react', 'react-hooks'],
  env: { browser: true },
  parserOptions: {
    ecmaVersion: 9,
    sourceType: 'module',
  },
  settings: {
    react: { version: 'detect' },
    // 'import/resolver': { node: { extensions: ['.js', '.tsx'], moduleDirectory: ['node_modules', './src'] } },
  },
  overrides: [
    {
      files: ['**/*.tsx'],
      rules: {
        'consistent-return': 'off', // clashes with useEffect,
        'react/prop-types': 'off',
      },
    },
  ],
  rules: {
    // base overrides
    'function-paren-newline': 'off',

    // proposed rules
    // 'react/boolean-prop-naming': 'off',
    // 'react/display-name': 'off',
    // 'react/jsx-key': 'off',
    // 'react/require-optimization': 'off',

    // to check
    // 'react/jsx-indent': ['error', 2],
    // 'indent': ['warn', 2, {SwitchCase: 1, ignoredNodes: ['JSXAttribute', 'JSXSpreadAttribute']}]
    // 'indent': [2, 4, {'SwitchCase': 1, 'VariableDeclarator': 1}],
    // 'react/prop-types': [2, { ignore: ['children'] }],

    // eslint
    'jsx-quotes': 'error',

    // disabled rules
    'react/jsx-filename-extension': 'off', // This will allow to fine tune tooling like a shared lint configuration
    'react/jsx-indent-props': 'off',
    'react/destructuring-assignment': 'off',
    'react/forbid-prop-types': 'off',
    'react/jsx-closing-tag-location': 'off',
    'react/jsx-curly-newline': 'off',
    'react/jsx-handler-names': 'off',
    'react/jsx-max-depth': 'off',
    'react/jsx-max-props-per-line': 'off',
    'react/jsx-no-literals': 'off',
    'react/jsx-no-target-blank': 'off',
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-props-no-spreading': 'off',
    'react/jsx-sort-default-props': 'off',
    'react/jsx-sort-props': 'off',
    'react/jsx-wrap-multilines': 'off',
    'react/no-multi-comp': 'off',
    'react/no-set-state': 'off',
    'react/require-default-props': 'off',
    'react/require-extension': 'off',
    'react/self-closing-comp': 'off',
    'react/sort-comp': 'off',
    'react/sort-prop-types': 'off',

    // jsx-a11y
    'jsx-a11y/alt-text': 'error',
    'jsx-a11y/anchor-has-content': 'error',
    'jsx-a11y/anchor-is-valid': ['error', { aspects: ['noHref', 'invalidHref'] }],
    'jsx-a11y/aria-activedescendant-has-tabindex': 'error',
    'jsx-a11y/aria-props': 'error',
    'jsx-a11y/aria-proptypes': 'error',
    'jsx-a11y/aria-role': ['error', { ignoreNonDOM: true }],
    'jsx-a11y/aria-unsupported-elements': 'error',
    'jsx-a11y/autocomplete-valid': 'error',
    'jsx-a11y/click-events-have-key-events': 'error',
    'jsx-a11y/heading-has-content': 'error',
    'jsx-a11y/html-has-lang': 'error',
    'jsx-a11y/iframe-has-title': 'error',
    'jsx-a11y/img-redundant-alt': 'error',
    'jsx-a11y/interactive-supports-focus': 'error',
    'jsx-a11y/label-has-associated-control': ['error', { controlComponents: ['Input'] }],
    'jsx-a11y/lang': 'error',
    'jsx-a11y/media-has-caption': 'error',
    'jsx-a11y/mouse-events-have-key-events': 'error',
    'jsx-a11y/no-access-key': 'error',
    'jsx-a11y/no-distracting-elements': 'error',
    'jsx-a11y/no-interactive-element-to-noninteractive-role': 'error',
    'jsx-a11y/no-noninteractive-element-interactions': 'error',
    'jsx-a11y/no-noninteractive-element-to-interactive-role': 'error',
    'jsx-a11y/no-noninteractive-tabindex': 'error',
    'jsx-a11y/no-onchange': 'error',
    'jsx-a11y/no-redundant-roles': 'error',
    'jsx-a11y/no-static-element-interactions': 'error',
    'jsx-a11y/role-has-required-aria-props': 'error',
    'jsx-a11y/role-supports-aria-props': 'error',
    'jsx-a11y/scope': 'error',
    'jsx-a11y/tabindex-no-positive': 'error',

    // react-hooks
    'react-hooks/exhaustive-deps': 'error',
    'react-hooks/rules-of-hooks': 'error',

    // rules
    'react/button-has-type': 'error',
    'react/forbid-foreign-prop-types': ['error', { allowInPropTypes: true }],
    'react/function-component-definition': ['error', { namedComponents: 'arrow-function', unnamedComponents: 'arrow-function' }],
    'react/jsx-closing-bracket-location': ['error', 'tag-aligned'],
    'react/jsx-curly-spacing': 'error',
    'react/jsx-first-prop-new-line': ['error', 'multiline-multiprop'],
    'react/jsx-fragments': ['error', 'element'],
    'react/jsx-indent': ['error', 2, { checkAttributes: true }],
    'react/jsx-no-comment-textnodes': 'error',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-undef': 'error',
    'react/jsx-no-useless-fragment': 'error',
    'react/jsx-pascal-case': ['error', { allowAllCaps: true, ignore: [] }],
    'react/jsx-tag-spacing': 'error',
    'react/jsx-uses-react': 'error',
    'react/jsx-uses-vars': 'error',
    'react/no-danger-with-children': 'error',
    'react/no-deprecated': 'error',
    'react/no-direct-mutation-state': 'error',
    'react/no-is-mounted': 'error',
    'react/no-typos': 'error',
    'react/prop-types': 'error',
    'react/react-in-jsx-scope': 'error',
    'react/require-render-return': 'error',
    'react/style-prop-object': 'error',
    'react/no-did-mount-set-state': 'error',
  },
};
