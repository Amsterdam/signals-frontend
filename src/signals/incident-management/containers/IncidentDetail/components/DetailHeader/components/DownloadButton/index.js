import React, { useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import Button  from 'components/Button';
import Spinner  from 'components/Spinner';

import useFetch from 'hooks/useFetch';

const DownloadButton = ({ label, url, filename }) => {
  const { get, isLoading, data } = useFetch();

  useEffect(() => {
    if (!data) return;

    if (navigator.msSaveOrOpenBlob) {
      navigator.msSaveOrOpenBlob(data, filename);
    } else {
      const href = global.URL.createObjectURL(data);
      const link = document.createElement('a');
      link.href = href;
      link.download = filename;
      document.body.appendChild(link);
      link.click();

      global.URL.revokeObjectURL(href);
      document.body.removeChild(link);
    }
  }, [data, filename]);

  const handleDownload = useCallback(() => {
    get(url, {}, { responseType: 'blob' });
  }, [get, url]);

  return (
    <Button
      disabled={isLoading}
      iconRight={isLoading && <Spinner />}
      iconSize={20}
      variant="application"
      data-testid="download-button"
      onClick={handleDownload}
    >
      {label}
    </Button>
  );
};

DownloadButton.propTypes = {
  url: PropTypes.string.isRequired,
  filename: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
};

export default DownloadButton;
