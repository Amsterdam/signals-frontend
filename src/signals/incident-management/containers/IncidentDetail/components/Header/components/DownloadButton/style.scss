@import "config.scss";

.download-button {
  &__link {
    height: 0;
    width: 0;
    opacity: 0;
  }
}
