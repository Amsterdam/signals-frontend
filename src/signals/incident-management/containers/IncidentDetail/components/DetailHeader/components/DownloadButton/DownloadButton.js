// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2019 - 2021 Gemeente Amsterdam
import { useCallback, useEffect } from 'react'

import PropTypes from 'prop-types'

import Button from 'components/Button'
import LoadingIndicator from 'components/LoadingIndicator'
import useFetch from 'hooks/useFetch'

const DownloadButton = ({ label, url, filename }) => {
  const { get, isLoading, data } = useFetch()

  useEffect(() => {
    if (!data) return

    if (navigator.msSaveOrOpenBlob) {
      navigator.msSaveOrOpenBlob(data, filename)
    } else {
      const href = global.URL.createObjectURL(data)
      const link = document.createElement('a')
      link.href = href
      link.download = filename
      document.body.appendChild(link)
      link.click()

      global.URL.revokeObjectURL(href)
      document.body.removeChild(link)
    }
    // Disabling linter; `filename` dependency triggers a download when changing the incident
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data])

  const handleDownload = useCallback(() => {
    get(url, {}, { responseType: 'blob' })
  }, [get, url])

  return (
    <Button
      type="button"
      variant="application"
      disabled={isLoading}
      iconRight={isLoading && <LoadingIndicator />}
      iconSize={20}
      data-testid="download-button"
      onClick={handleDownload}
    >
      {label}
    </Button>
  )
}

DownloadButton.propTypes = {
  url: PropTypes.string.isRequired,
  filename: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
}

export default DownloadButton
