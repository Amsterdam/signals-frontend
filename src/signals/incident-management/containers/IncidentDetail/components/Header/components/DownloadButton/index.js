/* eslint-disable jsx-a11y/href-no-hash,jsx-a11y/anchor-has-content */
import React from 'react';
import PropTypes from 'prop-types';
import 'whatwg-fetch';

import './style.scss';

function handleDownload(url, filename, accessToken) {
  const headers = {};

  if (accessToken) {
    headers.Authorization = `Bearer ${accessToken}`;
  }
  headers.Origin = window.location.origin;

  console.log('request', url, { method: 'GET',
    headers,
    responseType: 'blob'
  });

  fetch(url, {
    method: 'GET',
    headers,
    responseType: 'blob'
  }).then((response) => response.blob())
    .then((blob) => {
      console.log('BLOB', blob);
      const link = document.querySelector('.download-button__link');
      link.href = window.URL.createObjectURL(blob);
      link.click();
    });
}

const DownloadButton = ({ label, url, filename, accessToken }) => (
  <div className="download-button align-self-center">
    <button
      className="incident-detail__button"
      onClick={() => handleDownload(url, filename, accessToken)}
    >{label}</button>

    <a
      href="#"
      className="download-button__link"
      download={filename}
    ></a>
  </div>
);

DownloadButton.propTypes = {
  url: PropTypes.string.isRequired,
  filename: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  accessToken: PropTypes.string
};

export default DownloadButton;
