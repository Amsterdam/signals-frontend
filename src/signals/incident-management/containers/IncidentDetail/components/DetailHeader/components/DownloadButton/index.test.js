import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import DownloadButton from './index';

describe('<DownloadButton />', () => {
  let props;

  beforeEach(() => {
    props = {
      label: 'PDF',
      url: 'https://api.data.amsterdam.nl/signals/v1/private/signals/3077/pdf',
      filename: 'SIA melding 3077.pdf',
    };

    global.window.URL.createObjectURL = jest.fn();
    global.window.URL.revokeObjectURL = jest.fn();
  });

  describe('rendering', () => {
    it('should render correctly', () => {
      const { queryByTestId } = render(<DownloadButton {...props} />);
      const downloadButton = queryByTestId('download-button');

      expect(downloadButton).toHaveTextContent(/^PDF$/);
      expect(downloadButton.disabled).toEqual(false);
    });
  });

  describe('events', () => {
    beforeEach(() => {
      const res = new Response('{"hello":"world"}', {
        status: 200,
        headers: {
          'Content-type': 'application/blob',
        },
      });

      fetch.mockReturnValue(Promise.resolve(res));
    });

    it('should download document', async () => {
      const { queryByTestId, findByTestId } = render(<DownloadButton {...props} />);
      const downloadButton = queryByTestId('download-button');

      // suppress console error because of unimplemented navigation in JSDOM
      // @see {@link https://github.com/jsdom/jsdom/issues/2112}
      global.window.console.error = jest.fn();

      act(() => {
        fireEvent.click(downloadButton);
      });

      global.window.console.error.mockRestore();

      expect(downloadButton.disabled).toEqual(true);

      expect(fetch).toHaveBeenCalledWith(
        props.url,
        expect.objectContaining({
          method: 'GET',
          responseType: 'blob',
        })
      );

      await findByTestId('download-button');

      expect(downloadButton.disabled).toEqual(false);
    });

    it('should handle IE', async () => {
      global.window.navigator.msSaveOrOpenBlob = jest.fn();

      const { getByTestId, findByTestId } = render(<DownloadButton {...props} />);
      const downloadButton = getByTestId('download-button');

      expect(global.window.navigator.msSaveOrOpenBlob).not.toHaveBeenCalled();

      act(() => {
        fireEvent.click(downloadButton);
      });

      await findByTestId('download-button');

      expect(global.window.navigator.msSaveOrOpenBlob).toHaveBeenCalled();
    });
  });
});
