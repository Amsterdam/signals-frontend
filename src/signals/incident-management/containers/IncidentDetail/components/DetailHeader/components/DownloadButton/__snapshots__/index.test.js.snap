// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DownloadButton /> rendering should render correctly 1`] = `
<div
  className="download-button align-self-center"
>
  <button
    className="incident-detail__button"
    onClick={[Function]}
    type="button"
  >
    PDF
  </button>
</div>
`;

exports[`<DownloadButton /> rendering should render correctly when logged in 1`] = `
<div
  className="download-button align-self-center"
>
  <button
    className="incident-detail__button"
    onClick={[Function]}
    type="button"
  >
    PDF
  </button>
</div>
`;
